README.md

aoscx
=========

This Ansible network role provides a set of platform-dependent configuration
 management modules specifically designed for the AOS-CX network device.

Requirements
------------

* Python 3 or later
* Ansible 2.8.1 or later  
  * Ansible 2.10+ requires `ansible.netcommon` collection to be installed  
* Minimum supported AOS-CX firmware version 10.04.
* Enable REST on your AOS-CX device with the following commands:
    ```
    switch(config)# https-server rest access-mode read-write
    switch(config)# https-server vrf mgmt
    ```
* Install all Ansible requirements, with the following command:
    ```
    ansible-galaxy install -r requirements.yml
    ```
* Install all Python requirements with the following command:
    ```
    python3 -m pip install -r <collection_path>/requirements.txt
    ```
    
  * After installing the collection using the `ansible-galaxy` command you'll find the path to the collection in the following output:
    ```bash
    ubuntu-vm: $ ansible-galaxy collection install ~/arubanetworks-aoscx-3.0.0.tar.gz -f
    Process install dependency map
    Starting collection install process
    Skipping 'ansible.netcommon' as it is already installed
    Installing 'arubanetworks.aoscx:3.0.0' to '/home/administrator/.ansible/collections/ansible_collections/arubanetworks/aoscx'
    ```
    The requirements.txt file is located in the collection path. Based on the previous example, the command to install all Python3 requirements
    is as follows:
    ```
    python3 -m pip  install -r /home/administrator/.ansible/collections/ansible_collections/arubanetworks/aoscx/requirements.txt
    ```
* Install all Ansible requirements, with the following command:
    ```
    ansible-galaxy install -r <collection_path>/requirements.yml
    ```    
  
SSH/CLI Modules
---------------
* To use the SSH/CLI modules `aoscx_config` and `aoscx_command`, SSH access must
 be enabled on your AOS-CX device. It is enabled by default.
    * If necessary, re-enable SSH access on the device with the following command:
    ```
    switch(config)# ssh server vrf mgmt
    ```
* The control machine's `known_hosts` file must contain the target device's public key.
    * Alternatively, host key checking by the control machine may be disabled, although this is not recommended.
    * To disable host key checking modify the ansible.cfg file (default /etc/ansible/ansible.cfg) to include:
      `host_key_checking = false`
  
#### Limitations and Notes
* The default command timeout is 30 seconds. If a command takes more than 30
 seconds to execute, the task will time out.
	* If you regularly encounter the `command timeout triggered, timeout value
	 is 30 secs` error, consider setting the environment variable 
	`ANSIBLE_PERSISTENT_COMMAND_TIMEOUT` to a greater value. See Ansible documentation [here](https://docs.ansible.com/ansible/latest/network/user_guide/network_debug_troubleshooting.html).

Installation
------------

Through Github, use the following command. Use option `-f` to overwrite current role version:

```
ansible-galaxy install git+https://github.com/aruba/aoscx-ansible-role.git
```

Through Galaxy:

```
ansible-galaxy install arubanetworks.aoscx_role
```

Inventory Variables
-------------------

The variables that should be defined in your inventory for your AOS-CX host are:

* `ansible_host`: IP address of switch in `A.B.C.D` format. For IPv6 hosts use a string and enclose in square brackets E.G. `'[2001::1]'`.
* `ansible_user`: Username for switch in `plaintext` format  
* `ansible_password`: Password for switch in `plaintext` format
* `ansible_network_os`: Must always be set to `aoscx`
* `ansible_connection`: Set to `httpapi` to use REST API modules, to `network_cli` to use SSH/CLI modules and to `aoscx` to use pyaoscx modules
  * See [below](#pyaoscx-modules) for info on our new pyaoscx implementation of the AOS-CX Ansible modules that will be the standard moving forward
  * See [below](#using-both-rest-api-and-sshcli-modules-on-a-host) for info on using both REST API modules and SSH/CLI modules on a host
* `ansible_httpapi_use_ssl`: (Only required for REST API modules) Must always be `True` as AOS-CX uses port 443 for REST
* `ansible_httpapi_validate_certs`: (Only required for REST API modules) Set `True` or `False` depending on if Ansible should attempt to validate certificates
* `ansible_acx_no_proxy`: Set to `True` or `False` depending if Ansible should bypass environment proxies to connect to AOS-CX
* `ansible_aoscx_validate_certs`: Set to `True` or `False` depending if Ansible should bypass validating certificates to connect to AOS-CX. Only required when `ansible_connection` is set to `aoscx`
* `ansible_aoscx_use_proxy`: Set to `True` or `False` depending if Ansible should bypass environment proxies to connect to AOS-CX. Only required when `ansible_connection` is set to `aoscx`.



pyaoscx Modules
---------------
In an effort to make use of our recently updated Python SDK for AOS-CX [Pyaoscx](https://pypi.org/project/pyaoscx/) we've redesigned our Ansible integration by making use of pyaoscx for all REST-API based modules.   
**What does this mean if I've been using Ansible with AOS-CX REST API modules?**   
Our previous implementation will continue to function but will not be supported for future modules. That means you should and eventually have to update your [Ansible Inventory variables](https://github.com/aruba/aoscx-ansible-role#pyaoscx-modules-only) to specify the `ansible_network_os=aoscx` and additional variables as well as install the pyaoscx Python package using Python3 pip, **all playbooks will remain the same**:   
`pip3 install pyaoscx`  
The AOS-CX Ansible Role will automatically determine if you have pyaoscx installed and will use that method when the `ansible_network_os` is set to `aoscx`. If it's set to `httpapi` it will continue to use the previous implementation method.    

### Sample Inventories:

#### pyaoscx Modules Only:

##### INI

```INI
aoscx_1 ansible_host=10.0.0.1 ansible_user=admin ansible_password=password ansible_network_os=aoscx ansible_connection=aoscx ansible_aoscx_validate_certs=False ansible_aoscx_use_proxy=False
```

##### YAML

```yaml
all:
  hosts:
    aoscx_1:
      ansible_host: 10.0.0.1
      ansible_user: admin
      ansible_password: password
      ansible_network_os: aoscx
      ansible_connection: aoscx  # REST API via pyaoscx connection method
      ansible_aoscx_validate_certs: False
      ansible_aoscx_use_proxy: False

```


#### REST API Modules Only:

##### INI

```INI
aoscx_1 ansible_host=10.0.0.1 ansible_user=admin ansible_password=password ansible_network_os=aoscx ansible_connection=httpapi ansible_httpapi_validate_certs=False ansible_httpapi_use_ssl=True ansible_acx_no_proxy=True
```

##### YAML


```yaml
all:
  hosts:
    aoscx_1:
      ansible_host: 10.0.0.1
      ansible_user: admin
      ansible_password: password
      ansible_network_os: aoscx
      ansible_connection: httpapi  # REST API connection method
      ansible_httpapi_validate_certs: False
      ansible_httpapi_use_ssl: True
      ansible_acx_no_proxy: True
```

#### SSH/CLI Modules Only:

##### INI

```INI
aoscx_1 ansible_host=10.0.0.1 ansible_user=admin ansible_password=password ansible_network_os=aoscx ansible_connection=network_cli
```

##### YAML

```yaml
all:
  hosts:
    aoscx_1:
      ansible_host: 10.0.0.1
      ansible_user: admin
      ansible_password: password
      ansible_network_os: aoscx
      ansible_connection: network_cli  # SSH connection method
```

Example Playbooks
-----------------

### Including the Role

If role installed through [Github](https://github.com/aruba/aoscx-ansible-role)
set role to `aoscx-ansible-role`:

```yaml
- hosts: all
  roles:
    - role: aoscx-ansible-role
  vars:
    ansible_python_interpreter: /usr/bin/python3
  gather_facts: False     
  tasks:
  - name: Create L3 Interface 1/1/3
    aoscx_l3_interface:
      interface: 1/1/3
      description: Uplink_Interface
      ipv4: ['10.20.1.3/24']
      ipv6: ['2001:db8::1234/64']
```

If role installed through [Galaxy](https://galaxy.ansible.com/arubanetworks/aoscx_role)
set role to `arubanetworks.aoscx_role`:

```yaml
- hosts: all
  roles:
    - role: arubanetworks.aoscx_role
  vars:
    ansible_python_interpreter: /usr/bin/python3
  gather_facts: False     
  tasks:
  - name: Create L3 Interface 1/1/3
    aoscx_l3_interface:
      interface: 1/1/3
      description: Uplink_Interface
      ipv4: ['10.20.1.3/24']
      ipv6: ['2001:db8::1234/64']
```

Using Both REST API and SSH/CLI Modules on a Host
-------------------------------------------------

To use both REST API and SSH/CLI modules on the same host, 
you must create separate plays such 
that each play uses either only REST API modules or only SSH/CLI modules.
A play cannot mix and match REST API and SSH/CLI module calls.
In each play, `ansible_connection` must possess the appropriate value 
according to the modules used. 
If the play uses REST API modules, the value should be `httpapi`. 
If the play uses SSH/CLI modules, the value should be `network_cli`.
 
A recommended approach to successfully using both types of modules for a host
is as follows:
1. Set the host variables such that Ansible will connect to the host using REST API,
like seen [above](#rest-api-modules-only).
2. In the playbook, in each play wherein the SSH/CLI
modules are used, set the `ansible_connection` to `network_cli`. 

The inventory should look something like this:

```yaml
all:
  hosts:
    aoscx_1:
      ansible_host: 10.0.0.1
      ansible_user: admin
      ansible_password: password
      ansible_network_os: aoscx
      ansible_connection: httpapi  # REST API connection method
      ansible_httpapi_validate_certs: False
      ansible_httpapi_use_ssl: True
      ansible_acx_no_proxy: True
```

and the playbook like this (note how the second play, which uses the SSH/CLI module `aoscx_command`,
sets the `ansible_connection` value accordingly):

```yaml
- hosts: all
  roles: 
    - role: arubanetworks.aoscx_role
  vars:
     ansible_python_interpreter: /usr/bin/python3
  gather_facts: False     
  tasks:
  - name: Adding or Updating Banner
    aoscx_banner:
      banner_type: banner
      banner: "Hi!"

- hosts: all
  roles: 
    - role: arubanetworks.aoscx_role
  vars:
    ansible_connection: network_cli
  gather_facts: False    
  tasks:
    - name: Execute show run on the switch
      aoscx_command:
        commands: ['show run']
```


Contribution
-------
At Aruba Networks we're dedicated to ensuring the quality of our products, so if you find any
issues at all please open an issue on our [Github](https://github.com/aruba/aoscx-ansible-role) and we'll be sure to respond promptly!

For more contribution opportunities follow our guidelines outlined in our [CONTRIBUTING.md](https://github.com/aruba/aoscx-ansible-role/blob/master/CONTRIBUTING.md)

License
-------

Apache 2.0

Author Information
------------------
 - Madhusudan Pranav Venugopal (@madhusudan-pranav-venugopal)
 - Yang Liu (@yliu-aruba)
 - Tiffany Chiapuzio-Wong (@tchiapuziowong)
 - Derek Wang (@derekwangHPEAruba)
 - Daniel Alvarado Bonilla (@daniel-alvarado)



RELEASE-NOTES.md
# 3.0.0
* Major implementation changes that make use of [pyaoscx](https://pypi.org/project/pyaoscx/)
* New VSX module that utilized pyaoscx to configure VSX on AOS-CX switches

# 2.1.3
## Notable Changes
* Minor fix for static route module

# 2.1.1

## Notable Changes
* Adds logic for initial password handling
* This release includes new functionality in the connection plugin that handles if the device is in a default 
state that requires the admin user password to be set. If the admin password needs to be set the aoscx_role will 
set the user password to be what's defined in the Ansible inventory for ansible_password.

## Known Issues
* On 10.4 CX core and aggregation platforms (8400,8325,8320) REST API is not enabled for read/write access by default. The 
workaround to set the admin password on a default switch is to use the SSH modules 
[aoscx_command](https://github.com/aruba/aoscx-ansible-role/blob/master/docs/aoscx_command.md) or 
[aoscx_config](https://github.com/aruba/aoscx-ansible-role/blob/master/docs/aoscx_config.md) in a playbook and the connection module will handle setting the initial password.

# 2.1.0

## Notable Changes
* New Modules and Bug fixes
* This release includes new modules that allow for SSH/CLI commands and configuration. Refer to module documentation found in [docs/](https://github.com/aruba/aoscx-ansible-role/tree/master/docs).

## New Modules
* aoscx_command - This module connects to the CX device via SSH and allows CLI commands to be executed.
* aoscx_config - This module connects to the CX device via SSH and allows CLI configuration commands to be executed.
* aoscx_facts - This module used REST API to retrieve a subset of information from the CX device.

## Known Issues
* To use the aoscx_facts module, the device must be running firmware version 10.4 or higher.

# 2.0.0

## Notable Changes
* New Modules including 10.4 Module Compatibility Fixes
* This release includes new modules that allow for configuration management, firmware uploading, and system booting. Refer to module documentation found in [docs/](https://github.com/aruba/aoscx-ansible-role/tree/master/docs).

## New Modules
* aoscx_backup_config - This module downloads the configuration from the switch. It can also download the configuration to a remote TFTP server.
* aoscx_boot_firmware - This module boots the switch with the image present in the mentioned partition.
* aoscx_checkpoint - This module creates a new checkpoint or copies a config from the checkpoint.
* aoscx_upload_config - This module uploads a configuration onto the switch. It can also upload the configuration from a remote TFTP server.
* aoscx_upload_firmware - This module uploads a firmware onto the switch. It can also upload the firmware from HTTP server.

## Known Issues
* There is a known issue with the following modules on all platforms and 10.4 firmware version. Please only use these modules with 10.3 at this time: 
** aoscx_l3_interface
** aoscx_vlan_interface
* To upload firmware using HTTP server with the module aoscx_upload_firmware, the device must be running firmware version 10.4 or higher.

# 1.0.5

## Notable Changes
* 2.9 Fixes and formatting changes
* This allows the modules to be compatible with 2.9 and the formatting to be compliant with ansible-sanity.

# 1.0.4

## Notable Changes
* Bug Fixes L2 Interface
* fixes bugs with l2_interface module and trunk vlans

# 1.0.3

## Notable Changes
* Minor Bug Fixes

# 1.0.2

## Notable Changes
* Minor Bug Fixes
* Modifies modules to fix error messages and bugs with L3 interfaces and VRFs.

# 1.0.1

## Notable Changes
* It has a few fixes for ansible-lint warnings. It also fixes the role name in example playbook in the README.

# 1.0.0

## Notable Changes
* This is the initial release for the AOS-CX Ansible Role.
* For this release, only 10.3 AOS-CX firmware version is supported.

library/aoscx_config.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# (C) Copyright 2020 Hewlett Packard Enterprise Development LP.
# GNU General Public License v3.0+
# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)


from __future__ import (absolute_import, division, print_function)
__metaclass__ = type


ANSIBLE_METADATA = {
    'metadata_version': '1.1',
    'status': ['preview'],
    'supported_by': 'certified'
}

DOCUMENTATION = '''
---
module: aoscx_config
version_added: "2.9"
short_description: Logs in and executes configuration commands on AOS-CX device via SSH connection
description:
  - This module allows configuration of running-configs on AOS-CX devices via SSH connection
author: Aruba Networks (@ArubaNetworks)
options:

  lines:
    description:
      - List of configuration commands to be executed. If "parents" is specified, these
        are the child lines contained under/within the parent entry. If "parents" is not 
        specified, these lines will be checked and/or placed under the global config level. 
        These commands must correspond with what would be found in the device's running-config.
    required: False
    type: list
    aliases: ['commands']

  parents:
    description:
      - Parent lines that identify the configuration section or context under which the
        "lines" lines should be checked and/or placed. 
    required: False
    type: list

  src:
    description:
      - Path to the file containing the configuration to load into the device.  The path can 
        be either a full system path to the configuration file if the value starts with "/"
        or a path relative to the directory containing the playbook. This argument is mutually 
        exclusive with the "lines" and "parents" arguments. This src file must have same 
        indentation as a live switch config. The operation is purely additive, as it doesn't remove
        any lines that are present in the existing running-config, but not in the source config.
    required: False
    type: path

  before:
    description:
      - Commands to be executed prior to execution of the parent and child lines. This option
        can be used to guarantee idempotency.
    required: False
    type: list

  after:
    description:
      - Commands to be executed following the execution of the parent and child lines. This 
        option can be used to guarantee idempotency.
    required: False
    type: list

  match:
    description:
      - Specifies the method of matching. Matching is the comparison against the existing 
        running-config to determine whether changes need to be applied.
        If "match" is set to "line," commands are matched line by line.  
        If "match" is set to "strict," command lines are matched with respect to position.  
        If "match" is set to "exact," command lines must be an equal match.  
        If "match" is set to "none," the module will not attempt to compare the source 
        configuration with the running-config on the remote device.
    default: line
    choices: ['line', 'strict', 'exact', 'none']
    required: False
    type: str

  replace:
    description:
      - Specifies the approach the module will take when performing configuration on the 
        device. 
        If "replace" is set to "line," then only the differing and missing configuration lines 
        are pushed to the device. 
        If "replace" is set to "block," then the entire command block is pushed to the device 
        if there is any differing or missing line at all.
    default: line
    choices: ['line', 'block']
    required: False
    type: str

  backup:
    description:
      - Specifies whether a full backup of the existing running-config on the device will be 
        performed before any changes are potentially made. If the "backup_options" value is not 
        specified, the backup file is written to the "backup" folder in the playbook root 
        directory. If the directory does not exist, it is created.
    required: False
    type: bool
    default: False

  backup_options:
    description:
      - File path and name options for backing up the existing running-config. 
        To be used with "backup."
    suboptions:
      filename:
        description:
          - Name of file in which the running-config will be saved.
        required: False
        type: str
      dir_path:
        description:
          - Path to directory in which the backup file should reside.
        required: False
        type: path
    type: dict

  running_config:
    description:
      - Specifies an alternative running-config to be used as the base config for matching. The 
        module, by default, will connect to the device and retrieve the current running-config 
        to use as the basis for comparison against the source.  This argument is handy for times 
        when it is not desirable to have the task get the current running-config, and instead use
        another config for matching. 
    aliases: ['config']
    required: False
    type: str

  save_when:
    description:
      - Specifies when to copy the running-config to the startup-config. When changes are made to 
        the device running-configuration, the changes are not copied to non-volatile storage by default.  
        If "save_when" is set to "always," the running-config will unconditionally be copied to 
        startup-config.
        If "save_when" is set to "never," the running-config will never be copied to startup-config.
        If "save_when" is set to "modified," the running-config will be copied to startup-config 
        if the two differ.
        If "save_when" is set to "changed," the running-config will be copied to startup-config 
        if the task modified the running-config.
    default: never
    choices: ['always', 'never', 'modified', 'changed']
    required: False
    type: str

  diff_against:
    description:
      - When using the "ansible-playbook --diff" command line argument this module can generate 
        diffs against different sources. This argument specifies the particular config against 
        which a diff of the running-config will be performed. 
        If "diff_against" is set to "startup," the module will return the diff of the running-config 
        against the startup configuration.
        If "diff_against" is set to "intended," the module will return the diff of the running-config 
        against the configuration provided in the "intended_config" argument.
        If "diff_against" is set to "running," the module will return before and after diff of the 
        running-config with respect to any changes made to the device configuration.
    choices: ['startup', 'intended', 'running']
    required: False
    type: str

  diff_ignore_lines:
    description:
      - Specifies one or more lines that should be ignored during the diff. This is used to 
        ignore lines in the configuration that are automatically updated by the system. This 
        argument takes a list of regular expressions or exact commands.
    required: False
    type: list

  intended_config:
    description:
      - Path to file containing the intended configuration that the device should conform to, and 
        that is used to check the final running-config against. To be used with "diff_against," 
        which should be set to "intended."
    required: False
    type: str

  provider:
    description: A dict object containing connection details.
    suboptions:
      auth_pass:
        description:
          - Specifies the password to use if required to enter privileged mode on the
            remote device. If authorize is false, then this argument does nothing.
            If the value is not specified in the task, the value of  environment variable
            ANSIBLE_NET_AUTH_PASS will be used instead.
        type: str
      authorize:
        description:
          - Instructs the module to enter privileged mode on the remote device before
            sending any commands. If not specified, the device will attempt to execute
            all commands in non-privileged mode. If the value is not specified in the
            task, the value of environment variable ANSIBLE_NET_AUTHORIZE will be used instead.
        type: bool
      host:
        description:
          - Specifies the DNS host name or address for connecting to the remote device over the
            specified transport. The value of host is used as the destination address for the transport.
        required: True
        type: str
      password:
        description:
          - Specifies the password to use to authenticate the connection to the remote device.
            This value is used to authenticate the SSH session. If the value is not specified
            in the task, the value of environment variable ANSIBLE_NET_PASSWORD will be used instead.
        type: str
      port:
        description:
          - Specifies the port to use when building the connection to the remote device.
        type: int
      ssh_keyfile:
        description:
          - Specifies the SSH key to use to authenticate the connection to the remote device.
            This value is the path to the key used to authenticate the SSH session. If the value
            is not specified in the task, the value of environment variable ANSIBLE_NET_SSH_KEYFILE
            will be used instead.
        type: path
      timeout:
        description:
          - Specifies the timeout in seconds for communicating with the network device for either
            connecting or sending commands. If the timeout is exceeded before the operation is completed,
            the module will error.
        type: int
      username:
        description:
          - Configures the username to use to authenticate the connection to the remote device.
            This value is used to authenticate the SSH session. If the value is not specified in the task,
            the value of environment variable ANSIBLE_NET_USERNAME will be used instead.
        type: str
    type: dict
'''  # NOQA

EXAMPLES = '''
- name: First delete VLAN 44, then configure VLAN 45, and lastly create VLAN 46
  aoscx_config:
    before: 
      - no vlan 44
    parents: 
      - vlan 45
    lines:
      - name testvlan
      - description test_vlan
    after: 
      - vlan 46

- name: Back up running-config, then create VLAN 100, and save running-config to startup-config if change was made
  aoscx_config:
    backup: True
    lines: 
      - vlan 100
    backup_options:
      filename: backup.cfg
      dir_path: /users/Home/
    save_when: changed

- name: Compare running-config with saved config
  aoscx_config:
    diff_against: intended
    intended_config: /users/Home/backup.cfg

- name: Configure VLAN 2345 and compare resulting running-config with previous running-config
  aoscx_config:
    lines: 
      - vlan 2345
    diff_against: running

- name: Upload a config from local system file onto device
  aoscx_config:
    src:  /users/Home/golden.cfg

- name: Update interface 1/1/4, matching only if both "parents" and "lines" are present
  aoscx_config:
    lines:
      - ip address 4.4.4.5/24
    parents: interface 1/1/4
    match: strict

- name: Configure a multi-line banner
  aoscx_config:
    lines:
      - hello this is a banner_motd
      - this is banner line 2 banner_motd
      - this is banner line 3 banner_motd
    before: "banner motd `"
    after: "`"
'''  # NOQA

RETURN = r''' # '''

from ansible.module_utils.aoscx import run_commands, get_config, load_config
from ansible.module_utils.aoscx import aoscx_argument_spec
from ansible.module_utils.aoscx import check_args as aoscx_check_args
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.network.common.config import NetworkConfig, dumps


def get_running_config(module, config=None):
    '''
    Gets the running-config from the switch
    '''
    contents = module.params['running_config']
    if not contents:
        if config:
            contents = config
        else:
            contents = get_config(module)
    return NetworkConfig(contents=contents)


def get_candidate(module):
    '''
    Gets config candidate
    '''
    candidate = NetworkConfig()

    if module.params['src']:
        candidate.loadfp(module.params['src'])
    elif module.params['lines']:
        parents = module.params['parents'] or list()
        candidate.add(module.params['lines'], parents=parents)
    return candidate


def save_config(module, result):
    '''
    Saves config to memory
    '''
    result['changed'] = True
    if not module.check_mode:
        run_commands(module, 'write memory')
    else:
        module.warn('Skipping command `write memory` '
                    'due to check_mode.  Configuration not copied to '
                    'non-volatile storage')


def main():
    """ main entry point for module execution
    """
    backup_spec = dict(
        filename=dict(),
        dir_path=dict(type='path')
    )

    argument_spec = dict(
        src=dict(type='path'),

        lines=dict(aliases=['commands'], type='list'),
        parents=dict(type='list'),

        before=dict(type='list'),
        after=dict(type='list'),

        match=dict(default='line',
                   choices=['line', 'strict', 'exact', 'none']),
        replace=dict(default='line', choices=['line', 'block']),

        running_config=dict(aliases=['config']),
        intended_config=dict(),

        backup=dict(type='bool', default=False),
        backup_options=dict(type='dict', options=backup_spec),

        save_when=dict(choices=['always', 'never', 'modified', 'changed'],
                       default='never'),

        diff_against=dict(choices=['running', 'startup', 'intended']),
        diff_ignore_lines=dict(type='list'),
    )

    argument_spec.update(aoscx_argument_spec)

    mutually_exclusive = [('lines', 'src'),
                          ('parents', 'src')]

    required_if = [('match', 'strict', ['lines']),
                   ('match', 'exact', ['lines']),
                   ('replace', 'block', ['lines']),
                   ('diff_against', 'intended', ['intended_config'])]

    module = AnsibleModule(argument_spec=argument_spec,
                           mutually_exclusive=mutually_exclusive,
                           required_if=required_if,
                           supports_check_mode=True)

    warnings = list()
    aoscx_check_args(module, warnings)
    result = {'changed': False, 'warnings': warnings}

    config = None

    if module.params['diff_against'] is not None:
        module._diff = True

    if module.params['backup'] or (module._diff and
                                   module.params['diff_against'] == 'running'):
        contents = get_config(module)
        config = NetworkConfig(contents=contents)
        if module.params['backup']:
            result['__backup__'] = contents
            result['backup_options'] = module.params['backup_options']
            if module.params['backup_options']:
                if 'dir_path' in module.params['backup_options']:
                    dir_path = module.params['backup_options']['dir_path']
                else:
                    dir_path = ""
                if 'filename' in module.params['backup_options']:
                    filename = module.params['backup_options']['filename']
                else:
                    filename = "backup.cfg"

                with open(dir_path + '/' + filename, 'w') as backupfile:
                    backupfile.write(contents)
                    backupfile.write("\n")

    if any((module.params['src'], module.params['lines'])):
        match = module.params['match']
        replace = module.params['replace']

        candidate = get_candidate(module)

        if match != 'none':
            config = get_running_config(module, config)
            path = module.params['parents']
            configobjs = candidate.difference(
                config, match=match, replace=replace, path=path)
        else:
            configobjs = candidate.items

        if configobjs:
            commands = dumps(configobjs, 'commands').split('\n')

            if module.params['before']:
                commands[:0] = module.params['before']

            if module.params['after']:
                commands.extend(module.params['after'])

            result['commands'] = commands
            result['updates'] = commands

            if not module.check_mode:
                load_config(module, commands)

            result['changed'] = True

    running_config = None
    startup_config = None

    diff_ignore_lines = module.params['diff_ignore_lines']
    if diff_ignore_lines is None:
        diff_ignore_lines = []

    diff_ignore_lines.append("Current configuration:")
    diff_ignore_lines.append("Startup configuration:")

    if module.params['save_when'] == 'always':
        save_config(module, result)
    elif module.params['save_when'] == 'modified':
        output = run_commands(module,
                              ['show running-config', 'show startup-config'])

        running_config = NetworkConfig(
            contents=output[0], ignore_lines=diff_ignore_lines)
        startup_config = NetworkConfig(
            contents=output[1], ignore_lines=diff_ignore_lines)

        if running_config.sha1 != startup_config.sha1:
            save_config(module, result)
    elif module.params['save_when'] == 'changed':
        if result['changed']:
            save_config(module, result)

    if module._diff:
        if not running_config:
            output = run_commands(module, 'show running-config')
            contents = output[0]
        else:
            contents = running_config.config_text

        # recreate the object in order to process diff_ignore_lines
        running_config = NetworkConfig(
            contents=contents, ignore_lines=diff_ignore_lines)

        if module.params['diff_against'] == 'running':
            if module.check_mode:
                module.warn("unable to perform diff against "
                            "running-config due to check mode")
                contents = None
            else:
                contents = config.config_text

        elif module.params['diff_against'] == 'startup':
            if not startup_config:
                output = run_commands(module, 'show startup-config')
                contents = output[0]
            else:
                contents = startup_config.config_text

        elif module.params['diff_against'] == 'intended':
            with open(module.params['intended_config'], 'r') as intended_file:
                contents = intended_file.read()
        if contents is not None:
            base_config = NetworkConfig(
                contents=contents, ignore_lines=diff_ignore_lines)

            if running_config.sha1 != base_config.sha1:
                result.update({
                    'changed': True,
                    'diff': {'before': str(base_config),
                             'after': str(running_config)}
                })

    module.exit_json(**result)


if __name__ == '__main__':
    main()
